vllm/distributed/device_communicators/custom_all_reduce_utils.py

```diff
@@ -0,0 +1,165 @@
+import json
+import os
+import sys
+from contextlib import contextmanager
+from typing import Callable, Dict, List, Optional
+
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+
+import vllm.envs as envs
+from vllm.distributed.parallel_state import get_cpu_world_group, get_local_rank
+from vllm.logger import init_logger
+
+logger = init_logger(__name__)
+
+
+@contextmanager
+def mute_output():
+    with open(os.devnull, "w") as f:
+        sys.stderr = f
+        sys.stdout = f
+        yield
+
+
+def producer(i, CUDA_VISIBLE_DEVICES=None):
+    if CUDA_VISIBLE_DEVICES is not None:
+        os.environ["CUDA_VISIBLE_DEVICES"] = CUDA_VISIBLE_DEVICES
+    with mute_output():
+        dist.init_process_group(
+            backend="gloo",
+            init_method="tcp://127.0.0.1:12345",
```

Comment:
ideally we should. but we cannot use `get_open_port`. If `VLLM_PORT` is set, it will only return one value, which is already used for distributed communication.

Reply:
fixed in [34900d3](https://github.com/vllm-project/vllm/pull/4992/commits/34900d3548120990391862b31be2c6f1ec59d385)